lib/docker/cli/image.go

```diff
@@ -93,6 +88,79 @@ func (tarer DefaultImageTarer) CreateTarReader(imageName image.Name) (io.Reader,
 	return os.Open(targetPath)
 }
 
+// WriteTar imports an image, as a tar, to the image store
+func (tarer DefaultImageTarer) WriteTar(imageName image.Name, tarPath string) error {
+	repo, tag := imageName.GetRepository(), imageName.GetTag()
+
+	// Extract tar into temporary directory
+	dir := filepath.Join(tarer.store.SandboxDir, repo, tag)
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return fmt.Errorf("create unpack directory: %s", err)
+	}
+	defer os.RemoveAll(dir)
+
+	if err := snapshot.CreateDirectoryFromTar(dir, tarPath); err != nil {  // TODO
+		return fmt.Errorf("unpack tar: %s", err)
+	}
+
+	// Read manifest
+	exportManifestPath := filepath.Join(dir, "manifest.json")
+	var exportManifests []image.ExportManifest
+
+	if exportManifestJSON, err := ioutil.ReadFile(exportManifestPath); err != nil {
+		return fmt.Errorf("read manifest: %s", err)
+	} else if err := json.Unmarshal(exportManifestJSON, exportManifests); err != nil {
+		return fmt.Errorf("unmarshal manifest: %s", err)
+	}
+
+	for _, exportManifest := range exportManifests {
+
+		// Import extracted dir content into image store -- manifest.json
+		distManifest, err := image.NewDistributionManifestFromExport(exportManifest, dir)
+		if err != nil {
+			return fmt.Errorf("create distribution manifest: %s", err)
+		}
+		distManifestJSON, err := json.Marshal(distManifest)
+		if err != nil {
+			return fmt.Errorf("marshal manifest to JSON: %s", err)
+		}
+
+		distManifestFile, err := ioutil.TempFile(tarer.store.SandboxDir, "")
+		if err != nil {
+			return fmt.Errorf("create tmp manifest file: %s", err)
+		}
+		if _, err := distManifestFile.Write(distManifestJSON); err != nil {
+			return fmt.Errorf("write manifest file: %s", err)
+		}
+		if err := distManifestFile.Close(); err != nil {
+			return fmt.Errorf("close manifest file: %s", err)
+		}
+
+		distManifestPath := distManifestFile.Name()
+		if err = tarer.store.Manifests.LinkStoreFileFrom(repo, tag, distManifestPath); err != nil && !os.IsExist(err) {
+			return fmt.Errorf("commit manifest to store: %s", err)
+		}
+
+		// Import extracted dir content into image store -- {sha}.json
+		configPath := filepath.Join(dir, exportManifest.Config.String())
+		configID := exportManifest.Config.ID()
+		if err = tarer.store.Layers.LinkStoreFileFrom(configID, configPath); err != nil && !os.IsExist(err) {
+			return fmt.Errorf("commit config to store: %s", err)
+		}
+
+		// Import extracted dir content into image store -- {sha}/layer.tar
+		for _, layer := range exportManifest.Layers {
+			layerPath := path.Join(dir, layer.String())
+			layerID := layer.ID()
```

Comment:
if the export manifest is from `docker save`, the ID/SHA would actually be random (though the ordering would be correct). Need to fix that.